ORCA should generate plans with the ProjectSet node

DevChattopadhyay opened the pull request:
Motivation:
=========== 
In postgres 11 a new executor node ProjectSet was introduced that can evaluate
targetlists containing one or more set returning functions. Refer commit
69f4b9c. This was absorbed in ORCA in Postgres12 merge.

Query: `explain select generate_series(1,3)+1 as x;` 
Currently for the above query ORCA will generate a Result dxl node containing
set returning function in the projection list(target list). But since after
postgres11, ProjectSet nodes are only allowed to evaluate targetlists
containing set returning functions, this query will fallback to planner. This
PR is to implement ProjectSet node within ORCA to handle set returning
functions in the targetlist.

Implementation:
=============== 
The approach taken in this PR is to use the targetlist of the Result dxl node.
While translating the Result dxl node to planned statement node, the target
list of Result dxl node is checked for any set returning functions and if
present then required ProjectSet nodes are created using that. The advantage
of this approach is that we already have a targetlist and we just need to
figure out if ProjectSet nodes are required and create them. This all can be
done by using postgres functions. But the con is that it adds a lot of logic in
the translation part of Result dxl node.

Another Approaches:
=================== 
There can also be another or may be better approaches to
solve this problem.

We can create ProjectSet dxl nodes while translation of expression to dxl. But
the issue with this approach is that currently the CPhysicalComputeScalar
operator which is translated to a Result dxl node contains the set returning
functions in the Project list. So in order to create ProjectSet nodes from
CPhysicalComputeScalar we need to separate out all the set returning functions
which is tricky and have to create a lot of project list trees which will be
used to create multiple ProjectSet nodes. Also we need a logic to determine if
we need a Result dxl node on top of a ProjectSet node.

Another approach would be to create a CLogicalProjectSet expression to handle
set returning functions and traverse it through ORCA to generate a
CPhysicalProjectSet and create a dxl out of it. This will be the more
appropriate approach and will be done as a part of future work.

Before Fix:
============
- ORCA would currently fallback as soon as we see an SRF in the targetlist.

```
explain select generate_series(1,3)+1 as x;

LOG:  statement: explain select generate_series(1,3)+1 as x;
LOG:  2022-10-18 15:27:38:782047 IST,THD000,NOTICE,"Feature not supported: Unexpected target list entries in ProjectSet node",
                      QUERY PLAN
------------------------------------------------------
 Result  (cost=0.00..0.08 rows=3 width=4)
   ->  ProjectSet  (cost=0.00..0.03 rows=3 width=4)
         ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer: Postgres query optimizer
(4 rows)
```

After Fix:
===========
- ORCA is now able to generate a ProjectSet planned statement node.

```
explain select generate_series(1,4)+1 as x;
                      QUERY PLAN
------------------------------------------------------
 Result  (cost=0.00..0.00 rows=1 width=4)
   ->  ProjectSet  (cost=0.00..0.00 rows=1 width=4)
         ->  Result  (cost=0.00..0.00 rows=1 width=1)
 Optimizer: Pivotal Optimizer (GPORCA)
(4 rows)

select generate_series(1,4)+1 as x;
 x
---
 2
 3
 4
 5
(4 rows)
```

chrishajas commented:
I think I'm ok with the overall approach if we can refactor it and simplify the `TranslateDXLResult()` function. We're adding quite a bit of logic to the existing function, which I think makes it more difficult to understand. Would it be possible to have a check that if we'll need a ProjectSet node, we call a different function and separate the logic from the existing code?

dgkimura commented:
At the very least, can we leave a detailed paper trail for why we didn't/couldn't add a new node in optimizer to handle project set? I feel like the goal of translator is to be pretty dumb without too much logic, but this seems to go in the opposite direction. And if that's the case then my fear is that if we don't do the right thing now, then we're making it harder for ourselves later and we'll never do it. Thoughts?

DevChattopadhyay commented:
As discussed offline, we are currently going with the approach to create a Project Set plan node from a result dxl node for which this PR is for. The better approach will be to actually create a Clogical node to handle the set returning functions and then creating CPhysical, dxl and plan nodes.But this will take some effort and time.Will create another story for the creation of project set nodes using the better approach and will target it later.

dgkimura commented:
On Monday, May 15, 2023 10:09 AM Dev Swaroop Chattopadhyay wrote:
> Will create another story for the creation of project set nodes

Creating a story is good.  But I think we should probably document this somehow in the code, unless you're planning to pick that up right away.  That way it's not hidden somewhere only select people have access; and instead it's more accessible where anyone can quickly pick up.

I noticed this tripped assertion on your build:

```
test=# select generate_series(1,3)+1 as x from (select generate_series(1, 3)) as y;
FATAL:  Unexpected internal error (assert.c:44)
DETAIL:  FailedAssertion("!(hassrf)", File: "nodeProjectSet.c", Line: 196)
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
!> 
```

DevChattopadhyay commented:
> I noticed this tripped assertion on your build:
> 
> ```
> test=# select generate_series(1,3)+1 as x from (select generate_series(1, 3)) as y;
> FATAL:  Unexpected internal error (assert.c:44)
> DETAIL:  FailedAssertion("!(hassrf)", File: "nodeProjectSet.c", Line: 196)
> server closed the connection unexpectedly
>         This probably means the server terminated abnormally
>         before or while processing the request.
> The connection to the server was lost. Attempting reset: Failed.
> !> 
> ```

@dgkimura Thanks for pointing this out.I have fixed the code and added the above scenario in the testcase.I have also commented in the code the approach we are taking and what the better approach can be.Let me know if you want something else to be done.

dgkimura commented:
On your branch, I think this query's plan is bad:
```
test=# EXPLAIN (COSTS OFF) SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Result
   ->  ProjectSet
         ->  GroupAggregate
               Group Key: (true), (generate_series(1, 3)), (true)
               ->  Sort
                     Sort Key: (true), (generate_series(1, 3)), (true)
                     ->  Nested Loop Left Join
                           Join Filter: true
                           ->  ProjectSet
                                 ->  Result
                           ->  Materialize
                                 ->  Result
 Optimizer: Pivotal Optimizer (GPORCA)
 
test=# SET optimizer=on; SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
SET
 x 
---
 t
 f
 t
 f
 f
 t
(6 rows)

test=# SET optimizer=off; SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
SET
 x 
---
 t
 t
 t
(3 rows
```

gpopt commented:
> On your branch, I think this query's plan is bad:

Nice counter example.

DevChattopadhyay commented:
@dgkimura 

> On your branch, I think this query's plan is bad:

I checked this query on 6X and saw that its showing the same plan as 7X(only difference is 7X has project set nodes to contain SRF) and giving incorrect output.

```
show optimizer;
 optimizer
-----------
 on
(1 row)

explain verbose  SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
                                                                       QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------
 Result  (cost=0.00..441344.16 rows=1 width=1)
   Output: (((generate_series(1, 3)) = generate_series(1, (generate_series(1, 3)))) AND CASE WHEN (NOT (ColRef_0005 IS NULL)) THEN true ELSE false END)
   ->  GroupAggregate  (cost=0.00..441344.16 rows=1 width=5)
         Output: (generate_series(1, 3)), ColRef_0005,
         Group Key: "outer"., (generate_series(1, 3)), "outer".ColRef_0005
         ->  Sort  (cost=0.00..441344.16 rows=1 width=6)
               Output: , (generate_series(1, 3)), ColRef_0005
               Sort Key: "outer"., (generate_series(1, 3)), "outer".ColRef_0005
               ->  Nested Loop Left Join  (cost=0.00..441344.16 rows=1 width=6)
                     Output: , (generate_series(1, 3)), ColRef_0005
                     Join Filter: true
                     ->  Result  (cost=0.00..0.00 rows=1 width=5)
                           Output: generate_series(1, 3),
                           ->  Result  (cost=0.00..0.00 rows=1 width=1)
                                 Output: true
                     ->  Materialize  (cost=0.00..0.00 rows=1 width=1)
                           Output: ColRef_0005
                           ->  Result  (cost=0.00..0.00 rows=1 width=1)
                                 Output: true
                                 ->  Result  (cost=0.00..0.00 rows=1 width=1)
                                       Output: true
 Optimizer: Pivotal Optimizer (GPORCA)
 Settings: optimizer=on
(23 rows)

SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
 x
---
 t

 f
(3 rows)
```

DevChattopadhyay commented:
> On your branch, I think this query's plan is bad:

**On 6X_STABLE**

```
EXPLAIN SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
LOG:  statement: EXPLAIN SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
LOG:  2023-05-19 11:18:26:340451 IST,THD000,TRACE,"
Physical plan:
+--CPhysicalComputeScalar   rows:1   width:1  rebinds:1   cost:441344.162201   origin: [Grp:15, GrpExpr:14]
   |--CPhysicalStreamAgg( Global ) Grp Cols: ["" (0), "a" (1), "ColRef_0005" (5)], Minimal Grp Cols:["" (0), "a" (1), "ColRef_0005" (5)], Generates Duplicates :[ 0 ]    rows:2   width:6  rebinds:1   cost:441344.162200   origin: [Grp:22, GrpExpr:4]
   |  |--CPhysicalSort  ( (0.58.1.0), "" (0), NULLsLast ) ( (0.97.1.0), "a" (1), NULLsLast ) ( (0.58.1.0), "ColRef_0005" (5), NULLsLast )    rows:2   width:6  rebinds:1   cost:441344.162183   origin: [Grp:20, GrpExpr:4]
   |  |  +--CPhysicalLeftOuterNLJoin   rows:2   width:6  rebinds:1   cost:441344.162115   origin: [Grp:20, GrpExpr:3]
   |  |     |--CPhysicalComputeScalar   rows:1   width:5  rebinds:1   cost:0.000106   origin: [Grp:6, GrpExpr:1]
   |  |     |  |--CPhysicalConstTableGet Columns: ["" (0)] Values: [(1)]   rows:1   width:1  rebinds:1   cost:0.000001   origin: [Grp:0, GrpExpr:1]
   |  |     |  +--CScalarProjectList   origin: [Grp:5, GrpExpr:0]
   |  |     |     +--CScalarProjectElement "a" (1)   origin: [Grp:4, GrpExpr:0]
   |  |     |        +--CScalarFunc (generate_series)   origin: [Grp:3, GrpExpr:0]
   |  |     |           |--CScalarConst (1)   origin: [Grp:1, GrpExpr:0]
   |  |     |           +--CScalarConst (3)   origin: [Grp:2, GrpExpr:0]
   |  |     |--CPhysicalSpool (Streaming)   rows:1   width:1  rebinds:1   cost:0.000003   origin: [Grp:19, GrpExpr:2]
   |  |     |  +--CPhysicalComputeScalar   rows:1   width:1  rebinds:1   cost:0.000002   origin: [Grp:19, GrpExpr:1]
   |  |     |     |--CPhysicalConstTableGet Columns: ["" (2)] Values: [(1)]   rows:1   width:1  rebinds:1   cost:0.000001   origin: [Grp:10, GrpExpr:1]
   |  |     |     +--CScalarProjectList   origin: [Grp:18, GrpExpr:0]
   |  |     |        +--CScalarProjectElement "ColRef_0005" (5)   origin: [Grp:17, GrpExpr:0]
   |  |     |           +--CScalarConst (1)   origin: [Grp:16, GrpExpr:0]
   |  |     +--CScalarConst (1)   origin: [Grp:16, GrpExpr:0]
   |  +--CScalarProjectList   origin: [Grp:21, GrpExpr:0]
   +--CScalarProjectList   origin: [Grp:30, GrpExpr:0]
      +--CScalarProjectElement "x" (4)   origin: [Grp:29, GrpExpr:0]
         +--CScalarBoolOp (EboolopAnd)   origin: [Grp:28, GrpExpr:0]
            |--CScalarCmp (=)   origin: [Grp:9, GrpExpr:0]
            |  |--CScalarIdent "a" (1)   origin: [Grp:7, GrpExpr:0]
            |  +--CScalarFunc (generate_series)   origin: [Grp:8, GrpExpr:0]
            |     |--CScalarConst (1)   origin: [Grp:1, GrpExpr:0]
            |     +--CScalarIdent "a" (1)   origin: [Grp:7, GrpExpr:0]
            +--CScalarIf   origin: [Grp:27, GrpExpr:0]
               |--CScalarBoolOp (EboolopNot)   origin: [Grp:25, GrpExpr:0]
               |  +--CScalarNullTest   origin: [Grp:24, GrpExpr:0]
               |     +--CScalarIdent "ColRef_0005" (5)   origin: [Grp:23, GrpExpr:0]
               |--CScalarConst (1)   origin: [Grp:16, GrpExpr:0]
               +--CScalarConst (0)   origin: [Grp:26, GrpExpr:0]
```

**After this Fix on 7X:**

```
EXPLAIN SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
LOG:  statement: EXPLAIN SELECT a IN (SELECT generate_series(1,a)) AS x FROM (SELECT generate_series(1, 3) AS a) AS s;
LOG:  2023-05-19 11:14:40:590100 IST,THD000,TRACE,"
Physical plan:
+--CPhysicalComputeScalar   rows:1   width:1  rebinds:1   cost:441344.162201   origin: [Grp:15, GrpExpr:14]
   |--CPhysicalStreamAgg( Global ) Grp Cols: ["" (0), "a" (1), "ColRef_0005" (5)], Minimal Grp Cols:["" (0), "a" (1), "ColRef_0005" (5)], Generates Duplicates :[ 0 ]    rows:2   width:6  rebinds:1   cost:441344.162200   origin: [Grp:22, GrpExpr:4]
   |  |--CPhysicalSort  ( (0.58.1.0), "" (0), NULLsLast ) ( (0.97.1.0), "a" (1), NULLsLast ) ( (0.58.1.0), "ColRef_0005" (5), NULLsLast )    rows:2   width:6  rebinds:1   cost:441344.162183   origin: [Grp:20, GrpExpr:4]
   |  |  +--CPhysicalLeftOuterNLJoin   rows:2   width:6  rebinds:1   cost:441344.162115   origin: [Grp:20, GrpExpr:3]
   |  |     |--CPhysicalComputeScalar   rows:1   width:5  rebinds:1   cost:0.000106   origin: [Grp:6, GrpExpr:1]
   |  |     |  |--CPhysicalConstTableGet Columns: ["" (0)] Values: [(1)]   rows:1   width:1  rebinds:1   cost:0.000001   origin: [Grp:0, GrpExpr:1]
   |  |     |  +--CScalarProjectList   origin: [Grp:5, GrpExpr:0]
   |  |     |     +--CScalarProjectElement "a" (1)   origin: [Grp:4, GrpExpr:0]
   |  |     |        +--CScalarFunc (generate_series)   origin: [Grp:3, GrpExpr:0]
   |  |     |           |--CScalarConst (1)   origin: [Grp:1, GrpExpr:0]
   |  |     |           +--CScalarConst (3)   origin: [Grp:2, GrpExpr:0]
   |  |     |--CPhysicalSpool (Streaming)   rows:1   width:1  rebinds:1   cost:0.000003   origin: [Grp:19, GrpExpr:2]
   |  |     |  +--CPhysicalComputeScalar   rows:1   width:1  rebinds:1   cost:0.000002   origin: [Grp:19, GrpExpr:1]
   |  |     |     |--CPhysicalConstTableGet Columns: ["" (2)] Values: [(1)]   rows:1   width:1  rebinds:1   cost:0.000001   origin: [Grp:10, GrpExpr:1]
   |  |     |     +--CScalarProjectList   origin: [Grp:18, GrpExpr:0]
   |  |     |        +--CScalarProjectElement "ColRef_0005" (5)   origin: [Grp:17, GrpExpr:0]
   |  |     |           +--CScalarConst (1)   origin: [Grp:16, GrpExpr:0]
   |  |     +--CScalarConst (1)   origin: [Grp:16, GrpExpr:0]
   |  +--CScalarProjectList   origin: [Grp:21, GrpExpr:0]
   +--CScalarProjectList   origin: [Grp:30, GrpExpr:0]
      +--CScalarProjectElement "x" (4)   origin: [Grp:29, GrpExpr:0]
         +--CScalarBoolOp (EboolopAnd)   origin: [Grp:28, GrpExpr:0]
            |--CScalarCmp (=)   origin: [Grp:9, GrpExpr:0]
            |  |--CScalarIdent "a" (1)   origin: [Grp:7, GrpExpr:0]
            |  +--CScalarFunc (generate_series)   origin: [Grp:8, GrpExpr:0]
            |     |--CScalarConst (1)   origin: [Grp:1, GrpExpr:0]
            |     +--CScalarIdent "a" (1)   origin: [Grp:7, GrpExpr:0]
            +--CScalarIf   origin: [Grp:27, GrpExpr:0]
               |--CScalarBoolOp (EboolopNot)   origin: [Grp:25, GrpExpr:0]
               |  +--CScalarNullTest   origin: [Grp:24, GrpExpr:0]
               |     +--CScalarIdent "ColRef_0005" (5)   origin: [Grp:23, GrpExpr:0]
               |--CScalarConst (1)   origin: [Grp:16, GrpExpr:0]
               +--CScalarConst (0)   origin: [Grp:26, GrpExpr:0]
```

As shown above the Physical Plan is same on 6X_STABLE and after the fix and producing incorrect output. This PR contains translation from DXL to Planned Statement which happens after the Physical Plan is generated.So I think this issue is not related to this fix.Will create a seperate story for this issue.

gpopt commented on src/backend/optimizer/util/tlist.c:
```diff
@@ -1271,7 +1271,10 @@ split_pathtarget_at_srfs(PlannerInfo *root,
 				}
 			}
-			set_pathtarget_cost_width(root, ntarget);
+			if(NULL!=root)
```
This is the postgres style:
```
if (root != NULL)
```

Also add a comment this is specific for GPDB.

DevChattopadhyay replied:
Done

gpopt commented on src/include/gpopt/translate/CTranslatorDXLToPlStmt.h:
```diff
@@ -106,6 +106,24 @@ class CTranslatorDXLToPlStmt
+
+		Index m_newvarno;
+
+		int m_rtoffset;
```
Can you add comment to these struct and members?

DevChattopadhyay replied:
Done

pobbatihari commented:
Hii @DevChattopadhyay,
     While running queries on your branch, i got some  doubts on CTAS queries with SRF's.  I see the plan is same for both 6X and Master but i didn't understand why we have top result node.
     exp :
Master:
# explain create table x as select generate_series(1, 10)+1 as i;
                                       QUERY PLAN
----------------------------------------------------------------------------------------
 Result  (cost=0.00..0.00 rows=0 width=0)
   ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..0.01 rows=1 width=4)
         ->  Result  (cost=0.00..0.00 rows=1 width=4)
               One-Time Filter: (gp_execution_segment() = 0)
               ->  Result  (cost=0.00..0.00 rows=1 width=4)
                     ->  ProjectSet  (cost=0.00..0.00 rows=1 width=4)
                           ->  Result  (cost=0.00..0.00 rows=1 width=1)
6X:
explain create table x as select generate_series(1, 10)+1 as i;
                                       QUERY PLAN
----------------------------------------------------------------------------------------
 Result  (cost=0.00..0.00 rows=0 width=0)
   ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..0.01 rows=1 width=4)
         ->  Result  (cost=0.00..0.00 rows=1 width=8)
               ->  Result  (cost=0.00..0.00 rows=1 width=4)
                     One-Time Filter: (gp_execution_segment() = 0)
                     ->  Result  (cost=0.00..0.00 rows=1 width=4)
                           ->  Result  (cost=0.00..0.00 rows=1 width=1)

Master:
explain create table x as select generate_series(1, 10) as i;
                                       QUERY PLAN
----------------------------------------------------------------------------------------
 Result  (cost=0.00..0.00 rows=0 width=0)
   ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..0.01 rows=1 width=4)
         ->  Result  (cost=0.00..0.00 rows=1 width=4)
               One-Time Filter: (gp_execution_segment() = 0)
               ->  ProjectSet  (cost=0.00..0.00 rows=1 width=4)
                     ->  Result  (cost=0.00..0.00 rows=1 width=1)

6X:
explain create table x as select generate_series(1, 10) as i;
                                       QUERY PLAN
----------------------------------------------------------------------------------------
 Result  (cost=0.00..0.00 rows=0 width=0)
   ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..0.01 rows=1 width=4)
         ->  Result  (cost=0.00..0.00 rows=1 width=8)
               ->  Result  (cost=0.00..0.00 rows=1 width=4)
                     One-Time Filter: (gp_execution_segment() = 0)
                     ->  Result  (cost=0.00..0.00 rows=1 width=4)
                           ->  Result  (cost=0.00..0.00 rows=1 width=1)
  
 Can you please explain why do we need top result node with cost=0 rows=0 width=0 of above all queries.

gpopt commented on src/backend/optimizer/util/tlist.c:
```diff
@@ -1271,7 +1271,11 @@ split_pathtarget_at_srfs(PlannerInfo *root,
 			}
-			set_pathtarget_cost_width(root, ntarget);
+			/* This is specific for GPDB */
+			if(root != NULL)
```
Add space after `if`, this is the coding standard.

dgkimura commented on src/include/gpopt/translate/CTranslatorDXLToPlStmt.h:
```diff
@@ -591,6 +612,9 @@ class CTranslatorDXLToPlStmt
 		CDXLColRefArray *pdrgdxlcrOuterRefs, CDXLTranslateContext *dxltrctxLeft,
 		CDXLTranslateContext *dxltrctxRight);
+
+	static Node *fix_upper_expr_mutator_projectSet(
+		Node *node, fix_upper_expr_context_projectset *context);
```
[Nit] Why is this method format snake_case_style which is different than the others?  Could probably also use a quick comment.

DevChattopadhyay replied:
Modified to camel case

dgkimura commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3363,85 +3369,267 @@ CTranslatorDXLToPlStmt::TranslateDXLProjectSet(
+
+	ForEach(lc, targets_with_srf)
+	{
+		// The first element of the PathTarget list created by
+		// split_pathtarget_at_srfs method will not contain any
+		// SRF's. So skipping it.
+		if (list_cell_pos == 1)
+		{
+			list_cell_pos++;
+			continue;
+		}
```
Can you quickly explain why not?

DevChattopadhyay replied:
The logic is based on the description of the method `split_pathtarget_at_srfs`  which contains the line
`The first list entry must not contain any SRFs (other than ones duplicating input_target entries)`

dgkimura commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3271,79 +3318,38 @@ ContainsSetReturningFuncOrOp(const CDXLNode *project_list_dxlnode,
+CTranslatorDXLToPlStmt::TranslateDXLProjectSet(const CDXLNode *result_dxlnode)
 {
 	// GPDB_12_MERGE_FIXME: had we generated a DXLProjectSet in ORCA we wouldn't
 	// have needed to be defensive here...
```
Should this still be a FIXME or ORCA_FEATURE_NOT_SUPPORTED?  Is it still wrong to use one-time filter in ProjectSet node?  And if so, do we have a testcase?

dgkimura reviewed (commented):
Seems this is tricky problem, so it's awesome that you were able to get something working.  I'm still in-process of reviewing, but figured I'd post some thoughts now.

Rather than copying the Postgres commit message, I think we should instead explain why this particular PR is needed in Greenplum.  For example, we can describe the change in implementation and link to the commit 69f4b9c85f, acknowledge that was absorbed during Postgres 12 merge, explain how that effected ORCA and how this PR resolves that issue.  That is also probably a good place to document that we consciously chose to make this change in translator, the pros/cons for why we did that, and what it would entail to pursue alternative approaches (e.g. what part of the code needs to be updated).

Also, given that we found a query that generates wrong results, should this PR wait until we fix that?  Even if the issue is not directly related to this PR, it still exposes the issue by removing the fallback.  If we do choose to let this PR go in first, then I suggest either adding the testcase in this PR and/or creating a Github issue immediately after this PR goes in.  That way were are not hiding and won't forget the issue.

DevChattopadhyay commented on src/backend/gpopt/gpdbwrappers.cpp:
```diff
@@ -2561,4 +2561,83 @@ gpdb::GPDBLockRelationOid(Oid reloid, LOCKMODE lockmode)
+	GP_WRAP_START;
+	{
+		return split_pathtarget_at_srfs(root, target, input_target, targets,
+										targets_contain_srfs);
```
Removed return keyword

DevChattopadhyay commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3363,85 +3369,267 @@ CTranslatorDXLToPlStmt::TranslateDXLProjectSet(
+												 BOOL &will_require_result_node)
+{
+	// Method split_pathtarget_at_srfs will split the given PathTarget into multiple levels to position SRFs safely.
+	// This list will hold the splited PathTarget created by split_pathtarget_at_srfs method.
```
Done

DevChattopadhyay commented:
> On your branch, I think this query's plan is bad:

Opened a Github issue 
https://github.com/greenplum-db/gpdb/issues/15644

DevChattopadhyay commented:
> Can you please explain why do we need top result node with cost=0 rows=0 width=0 of above all queries.

Hi @pobbatihari 
I think there is already a opened github issue for this
https://github.com/greenplum-db/gpdb/issues/14629

DevChattopadhyay commented on src/include/gpopt/translate/CTranslatorDXLToPlStmt.h:
```diff
@@ -106,6 +106,24 @@ class CTranslatorDXLToPlStmt
+
+		// If a matching subplan output expression is found, a new VAR will be created
+		// with this varno(OUTER_VAR)
+		Index m_newvarno;
```
Removed the field.

dgkimura commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3396,85 +3404,262 @@ CTranslatorDXLToPlStmt::TranslateDXLProjectSet(
+		Var *newvar;
+
+		newvar = gpdb::MakeVarFromTargetEntry(newvarno, tle);
+		newvar->varnoold = 0;
+		newvar->varoattno = 0;
```
Why in this case do varnoold/varoattno need to be different than varno/varattno?

DevChattopadhyay replied:
I have borrowed this logic from Postgres code.Whenever a new VAR is created which reference the target list item of the subplan , varnoold and varoattno is set to 0 
```
/* Found a matching subplan output expression */
		Var		   *newvar;

		newvar = makeVarFromTargetEntry(newvarno, tle);
		newvar->varnoold = 0;	/* wasn't ever a plain Var */
		newvar->varoattno = 0;
		return newvar;
```

dgkimura replied:
I see. It's unfortunate that we need to duplicate that logic.  Would it be worth exposing it through gpdbwrappers?

DevChattopadhyay replied:
The logic is based on Postgres method 
```
search_indexed_tlist_for_non_var(Expr *node,
			indexed_tlist *itlist, Index newvarno)
```

This method takes `indexed_tlist` as an argument which has multiple fields that i don't require.In order to create this i need to call 
`build_tlist_index(List *tlist)`
which assigns memory through palloc which is a overhead.Due to these reasons i have not created a wrapper around it as it was a small function.

dgkimura reviewed (commented):
Few minor comments.  Otherwise, looks good to me.

DevChattopadhyay commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3293,9 +3293,56 @@ CTranslatorDXLToPlStmt::TranslateDXLSubQueryScan(
+//--------------------------------------------------------------------------------------
+// If the top level is not a function returning set then we need to check if the
+// project element contains any SRF's deep down the tree.If we found any SRF's at
+// lower levels then we will require a result node on top of ProjectSet node.Eg
```
Done

DevChattopadhyay commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3396,85 +3404,262 @@ CTranslatorDXLToPlStmt::TranslateDXLProjectSet(
-	// double check the targetlist is kosher
-	// we are only doing this because ORCA didn't do it...
-	if (!SanityCheckProjectSetTargetList(plan->targetlist))
+void
+CTranslatorDXLToPlStmt::MutateFuncExprToVarProjectSet(Plan *final_plan)
```
Added comment to the method

dgkimura replied:
Ty

DevChattopadhyay commented on src/test/regress/sql/qp_misc.sql:
```diff
@@ -15614,3 +15614,42 @@ INSERT INTO tparam VALUES ('a_value');
+------------------ Project Set Node  ------------------------
+-------------------------------------------------------------
+
+-- SRF is not under any other expression --
```
Done.Created gp_tsrf.sql

DevChattopadhyay commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3522,26 +3702,48 @@ CTranslatorDXLToPlStmt::TranslateDXLResult(
+	if (will_require_result_node)
+	{
+		result->plan.lefttree = project_set_parent_plan;
+		final_plan = &(result->plan);
```
final_plan dont have AddRef() method

dgkimura commented on src/include/gpopt/translate/CTranslatorDXLToPlStmt.h:
```diff
@@ -106,6 +106,20 @@ class CTranslatorDXLToPlStmt
 		}
 	};	// SContextIndexVarAttno
 
+	// struct that holds whatever context information the
+	// "FixUpperExprMutatorProjectSet" mutator routine needs.
+	struct FixUpperExprContextProjectset
+	{
+		// For a given node,this will contain the targetlist of the lefttree of that node
+		List *m_subplan_tlist;
+
+		FixUpperExprContextProjectset(List *subplan_tlist)
+			: m_subplan_tlist(subplan_tlist)
+		{
+		}
+
+	};	// FixUpperExprContextProjectset
```
Is there much value keeping this as its own struct now?  Seems we could probably just pass the list around wherever it was previously used?

dgkimura replied:
But technically a list is a struct..

DevChattopadhyay replied:
Oh yeah sorry.My bad.Removed the struct now.

dgkimura reviewed (approved):
A couple more minor things, but I don't want to block you from getting this in.  I would like to take a look at your final commit message and give it a read.  Otherwise, I think this looks good enough to iterate on later.

gpopt commented on src/backend/optimizer/util/tlist.c:
```diff
@@ -1271,7 +1271,15 @@ split_pathtarget_at_srfs(PlannerInfo *root,
-			set_pathtarget_cost_width(root, ntarget);
+
+			/*
+			 * This check is specific for ORCA. ORCA calls split_pathtarget_at_srfs function
+			 * for creation of ProjectSet nodes during planned statement translation with
+			 * PlannerInfo as NULL. So having this check prevents a crash in
+			 * set_pathtarget_cost_width.
```
Run command+GQ to format these line.

gpopt commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3396,85 +3404,262 @@ CTranslatorDXLToPlStmt::TranslateDXLProjectSet(
+	}
 
+	tle = gpdb::TlistMember(node, itlist);
+	if (tle)
```
!= nullptr

DevChattopadhyay replied:
Done

gpopt commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3522,26 +3702,48 @@ CTranslatorDXLToPlStmt::TranslateDXLResult(
+	return final_plan;
 }
 
+
```
remove this line change

gpopt commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3396,85 +3404,262 @@ CTranslatorDXLToPlStmt::TranslateDXLProjectSet(
-	// we are only doing this because ORCA didn't do it...
-	if (!SanityCheckProjectSetTargetList(plan->targetlist))
+	newvar = SearchTlistForNonVarProjectset((Expr *) node, context, OUTER_VAR);
+	if (newvar)
```
!= nullptr

DevChattopadhyay replied:
Done

gpopt reviewed (approved):
LGTM besides some cosmetic issues.
